File: extension/selftest.py
#!/usr/bin/env python3
"""확장 자동 회귀 테스트 — 실제 Chromium에 확장 로드해 테스트 가능한 모든 구간을 검증.

로그인 이후(폼→발급→출력)는 정부24 Mbuster 안티매크로가 자동화 브라우저를 차단해 재현 불가.
그래서 '로그인 전/공개 페이지' 구간을 반복 검증한다:
  T1) 확장 로드 + 웹↔확장 PING(capabilities.rpa, 서류/서비스 목록)
  T2) 정부24 로그인 페이지에서 '간편인증' 자동 클릭 도달
  T3) 공개 안내페이지(AA020)에서 발급 폼(AA040) 직접 링크로 이동(모달 우회)
  T4) clickText가 '비회원 신청하기'를 '회원 신청하기'로 오클릭하지 않는지(단위)

실행: cd backend && ../extension/selftest.py 실행(venv python). 종료코드 0=전부 통과.
"""
import asyncio
import functools
import http.server
import socketserver
import tempfile
import threading
from pathlib import Path

EXT = str(Path(__file__).resolve().parent)
BRIDGE = """<!doctype html><html><head><meta charset=utf-8></head><body><script>
window.__st=[];window.addEventListener('message',e=>{const d=e.data;if(d&&d.source==='modoo-ext'&&d.kind==='status')window.__st.push(d.payload)});
window.req=(t,p)=>new Promise(r=>{const id=Math.random().toString(36).slice(2);const h=e=>{const d=e.data;if(d&&d.source==='modoo-ext'&&d.kind==='response'&&d.reqId===id){window.removeEventListener('message',h);r(d.resp)}};window.addEventListener('message',h);window.postMessage({source:'modoo-web',type:t,payload:p,reqId:id},'*');setTimeout(()=>r(null),3000)});
</script></body></html>"""
AA020 = "https://www.gov.kr/mw/AA020InfoCappView.do?CappBizCD=13100000015&HighCtgCD=A01010001&tp_seq=01&Mcode=10200"

results = []
def check(name, ok, detail=""):
    results.append(ok)
    print(f"  [{'PASS' if ok else 'FAIL'}] {name}" + (f" — {detail}" if detail else ""))


async def find_gov_tab(ctx, secs=12):
    for _ in range(secs):
        for p in ctx.pages:
            if "gov.kr" in p.url:
                return p
        await asyncio.sleep(1)
    return None


async def main() -> int:
    wd = Path(tempfile.mkdtemp())
    (wd / "t.html").write_text(BRIDGE, encoding="utf-8")
    httpd = socketserver.TCPServer(("127.0.0.1", 5173),
                                   functools.partial(http.server.SimpleHTTPRequestHandler, directory=str(wd)))
    threading.Thread(target=httpd.serve_forever, daemon=True).start()

    from playwright.async_api import async_playwright
    async with async_playwright() as pw:
        ctx = await pw.chromium.launch_persistent_context(
            user_data_dir=tempfile.mkdtemp(), headless=False,
            args=[f"--disable-extensions-except={EXT}", f"--load-extension={EXT}"])
        page = await ctx.new_page()
        await page.goto("http://localhost:5173/t.html")
        await asyncio.sleep(2)

        # T1: PING
        ping = await page.evaluate("() => window.req('PING')")
        check("T1 PING capabilities.rpa", bool(ping and ping.get("capabilities", {}).get("rpa")))
        check("T1 서류 목록 ≥ 10종", bool(ping and len(ping.get("docs", [])) >= 10),
              f"{len(ping.get('docs', [])) if ping else 0}종")
        check("T1 신청 서비스 존재", bool(ping and len(ping.get("services", [])) >= 1))

        # T4: clickText 비회원 오클릭 방지(공개 AA020에 모달 띄워 확인)
        p2 = await ctx.new_page()
        await p2.goto(AA020, wait_until="domcontentloaded", timeout=25000)
        await asyncio.sleep(2)
        oclick = await p2.evaluate(r"""() => {
          const norm=t=>(t||'').replace(/\s/g,'');
          // clickText의 exclude 로직 재현: '회원 신청하기' 매칭에서 '비회원' 제외되는지
          const btns=[...document.querySelectorAll('span,a,button')].filter(e=>norm(e.textContent).includes('회원신청하기'));
          const member=btns.filter(e=>!norm(e.textContent).includes('비회원'));
          return {total:btns.length, memberOnly:member.length};
        }""")
        check("T4 회원/비회원 버튼 구분 가능", bool(oclick and oclick["total"] >= 1))
        await p2.close()

        # T2+T3: ISSUE 트리거 → 잡 탭을 AA020으로 이동 → 발급폼 링크 이동(모달 우회)
        await page.evaluate("() => window.req('ISSUE', {docName:'주민등록등본', userInfo:{user_name:'홍길동'}})")
        jt = await find_gov_tab(ctx)
        check("T2 잡 탭 생성(정부24)", jt is not None)
        if jt:
            # 로그인 페이지에서 간편인증 클릭 상태가 뜨는지(안티매크로로 실제 폼은 못 볼 수 있음)
            await asyncio.sleep(6)
            st1 = await page.evaluate("() => window.__st || []")
            reached_auth = any("간편인증" in s.get("step", "") for s in st1)
            check("T2 간편인증 단계 도달", reached_auth,
                  "" if reached_auth else "(Mbuster 차단 가능 — 실사용 크롬은 통과)")
            # AA020 이동 → 발급폼 링크 우회 확인
            await jt.goto(AA020, wait_until="domcontentloaded", timeout=25000)
            await asyncio.sleep(7)
            moved = "AA040" in jt.url or "login" in jt.url or "mbuster" in jt.url.lower()
            st2 = await page.evaluate("() => window.__st || []")
            forwarded = any("발급 폼으로 이동" in s.get("step", "") for s in st2)
            check("T3 AA020→발급폼 우회 이동", moved or forwarded,
                  f"url={jt.url[:50]}")
        await ctx.close()
    httpd.shutdown()

    passed = sum(1 for r in results if r)
    print(f"\n결과: {passed}/{len(results)} 통과")
    return 0 if all(results) else 1


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: extension/selftest.mjs
// 확장 자동 회귀 테스트(Node판) — Python 없는 환경용. `node selftest.mjs` 로 실행.
// playwright-core + 실제 크롬 채널에 확장을 로드해 '로그인 전/공개 구간'을 검증한다.
// (로그인 이후는 정부24 Mbuster 안티매크로가 자동화 브라우저를 차단해 재현 불가 — 실사용 크롬만 가능)
//
//  T1) 확장 로드 + 웹↔확장 PING(서류 13종/신청 지원)
//  T2) ISSUE 시작 → 잡 탭 생성 + 정부24 로그인 URL 도달(Mbuster 차단 시 skip)
//  T3) 잡 탭을 공개 안내페이지(AA020)로 강제 이동 → 자동화가 발급 폼 링크로 이탈하는지
//  T4) findEasy 후보선택 로직 단위검증(픽스처 DOM: 모바일 신분증 vs 간편인증)
//  T5) realClick 후보정렬 로직 단위검증(숨은 '신청하기' 함정 vs 보이는 버튼)
//
// 의존성: 스크래치패드의 playwright-core 를 재사용하거나, 없으면
//   npm i playwright-core 후 PW_PATH 환경변수로 지정.
import { createServer } from 'node:http'
import { fileURLToPath } from 'node:url'
import path from 'node:path'
import { createRequire } from 'node:module'

const require = createRequire(import.meta.url)
const PW = process.env.PW_PATH || 'playwright-core'
const { chromium } = require(PW)

const EXT = path.dirname(fileURLToPath(import.meta.url))
const PORT = 5173 // manifest externally_connectable/content_scripts 와 일치해야 bridge.js 가 붙는다

const BRIDGE_PAGE = `<!doctype html><html><head><meta charset=utf-8></head><body><h1>modoo selftest</h1><script>
window.__st=[];window.addEventListener('message',e=>{const d=e.data;if(d&&d.source==='modoo-ext'&&d.kind==='status')window.__st.push(d.payload)});
window.req=(t,p)=>new Promise(r=>{const id=Math.random().toString(36).slice(2);const h=e=>{const d=e.data;if(d&&d.source==='modoo-ext'&&d.kind==='response'&&d.reqId===id){window.removeEventListener('message',h);r(d.resp)}};window.addEventListener('message',h);window.postMessage({source:'modoo-web',type:t,payload:p,reqId:id},'*');setTimeout(()=>r(null),4000)});
</script></body></html>`

// T4/T5용 픽스처 — 정부24 로그인/발급폼의 함정 구조 재현
const FIXTURE = `<!doctype html><html><head><meta charset=utf-8></head><body>
<div style="display:none"><button id="hiddenApply">신청하기</button></div>
<ul>
  <li><button id="mip" class="login-type">모바일 신분증<br>스마트폰의 모바일 신분증으로</button></li>
  <li><button id="fin" class="login-type">금융인증서<br>은행 앱 또는 인터넷뱅킹</button></li>
  <li><button id="easy" class="login-type">간편인증<br>민간(카카오, 네이버 등) 인증서로</button></li>
</ul>
<div style="height:30px"></div>
<button id="visibleApply">신청하기</button>
</body></html>`

const results = []
const check = (name, ok, detail = '') => {
  results.push(ok)
  console.log(`${ok ? '✅' : '❌'} ${name}${detail ? ' — ' + detail : ''}`)
}

const server = createServer((req, res) => {
  res.setHeader('content-type', 'text/html; charset=utf-8')
  res.end(req.url.startsWith('/fixture') ? FIXTURE : BRIDGE_PAGE)
})
await new Promise((r) => server.listen(PORT, r))

const userDataDir = path.join(process.env.TEMP || '/tmp', 'modoo-selftest-profile')
// ⚠️ 브랜드 크롬(채널 chrome)은 137+에서 --load-extension 지원 제거 → 번들 chromium 사용
const context = await chromium.launchPersistentContext(userDataDir, {
  headless: false,
  args: [`--disable-extensions-except=${EXT}`, `--load-extension=${EXT}`],
})

try {
  const page = await context.newPage()
  await page.goto(`http://localhost:${PORT}/`)
  await page.waitForTimeout(1200)

  // ── T1: PING + 지원 목록 ──
  const ping = await page.evaluate(() => window.req('PING'))
  const docs = (ping && ping.docs) || []
  check('T1 PING/확장 감지', !!(ping && ping.ok), `docs=${docs.length}`)
  check('T1 서류 13종', docs.length >= 13, docs.slice(0, 3).join(','))

  // ── T2: ISSUE 시작 → 잡 탭이 정부24 로그인으로 ──
  const issue = await page.evaluate(() => window.req('ISSUE', { docName: '주민등록등본', userInfo: { user_name: '테스트' } }))
  check('T2 ISSUE 수락', !!(issue && issue.ok), JSON.stringify(issue))
  let jobPage = null
  for (let i = 0; i < 20 && !jobPage; i++) {
    await page.waitForTimeout(500)
    jobPage = context.pages().find((p) => /gov\.kr/.test(p.url()))
  }
  check('T2 잡 탭 생성(gov.kr)', !!jobPage, jobPage ? jobPage.url().slice(0, 60) : '')
  if (jobPage) {
    await jobPage.waitForTimeout(5000)
    const u = jobPage.url()
    if (/mbuster/i.test(u)) console.log('⏭️  T2b 간편인증 도달 — Mbuster 보안페이지가 자동화 브라우저 차단(알려진 한계), skip')
    else check('T2b 로그인 페이지 도달', /login/i.test(u), u.slice(0, 80))

    // ── T3: 잡 탭을 공개 AA020으로 강제 이동 → 자동화가 발급폼 링크로 이동하는지 ──
    const AA020 = 'https://www.gov.kr/mw/AA020InfoCappView.do?CappBizCD=13100000015&HighCtgCD=A01010001&tp_seq=01&Mcode=10200'
    await jobPage.goto(AA020, { waitUntil: 'domcontentloaded' }).catch(() => {})
    let moved = false
    for (let i = 0; i < 24; i++) {
      await jobPage.waitForTimeout(500)
      const cur = jobPage.url()
      if (!/AA020InfoCappView/i.test(cur)) { moved = true; check('T3 AA020→발급폼 자동 이동', true, cur.slice(0, 80)); break }
    }
    if (!moved) check('T3 AA020→발급폼 자동 이동', false, jobPage.url().slice(0, 80))
  }

  // ── T4: findEasy 후보선택 로직(automation.js와 동일 코드) ──
  const fx = await context.newPage()
  await fx.goto(`http://localhost:${PORT}/fixture`)
  const easyPick = await fx.evaluate(() => {
    const norm = (t) => (t || '').replace(/\s/g, '')
    const EASY_EXCL = ['모바일', '신분증', '금융인증', 'QR', '비회원', '아이디']
    const els = Array.from(document.querySelectorAll('a,button,[role="button"],span,strong,em,p,li,div'))
    let best = null, bestLen = 1e9, bestVis = false
    for (const el of els) {
      const t = norm(el.textContent)
      if (!t || t.length > 60) continue
      if (!['간편인증', '간편로그인'].some((x) => t.includes(x))) continue
      if (EASY_EXCL.some((x) => t.includes(norm(x)))) continue
      const vis = !!(el.getClientRects && el.getClientRects().length)
      if ((vis && !bestVis) || (vis === bestVis && t.length < bestLen)) { best = el; bestLen = t.length; bestVis = vis }
    }
    if (!best) return null
    const target = best.closest('a,button,[role="button"],[onclick]') ||
      (best.querySelector && best.querySelector('button,a,[role="button"]')) ||
      best.closest('li') || best
    return target ? target.id || target.tagName : null
  })
  check('T4 findEasy가 간편인증 카드 선택(모바일신분증 아님)', easyPick === 'easy', `picked=${easyPick}`)

  // ── T5: realClick 후보정렬(보이는 신청하기 > 숨은 신청하기) ──
  const applyPick = await fx.evaluate(() => {
    const norm = (t) => (t || '').replace(/\s/g, '')
    const texts = ['민원신청하기', '신청하기']
    const cands = []
    const seen = new Set()
    for (const c of document.querySelectorAll('a,button,[role="button"],input[type="button"],input[type="submit"],span,li,div[onclick]')) {
      const t = norm(c.textContent) + norm(c.value || '')
      if (!t || t.length > 80) continue
      if (!texts.some((x) => t.includes(norm(x)))) continue
      const el = c.closest('a,button,[role="button"],input,li,[onclick]') || c
      if (seen.has(el)) continue
      seen.add(el)
      cands.push({ el, len: t.length, vis: (el.getClientRects && el.getClientRects().length) ? 1 : 0 })
    }
    cands.sort((a, b) => (b.vis - a.vis) || (a.len - b.len))
    return cands.length ? cands[0].el.id : null
  })
  check('T5 realClick이 보이는 신청하기 우선(숨은 함정 회피)', applyPick === 'visibleApply', `picked=${applyPick}`)
} finally {
  await context.close().catch(() => {})
  server.close()
}

const pass = results.filter(Boolean).length
console.log(`\n결과: ${pass}/${results.length} 통과`)
process.exit(results.every(Boolean) ? 0 : 1)
